### PR TITLE
DAOS-8905 control: Add sysfs net device detection

### PR DESCRIPTION
**Pull request by kjacque**
- Add a TopologyProvider to detect network devices via sysfs, to
      find devices not in the hwloc topology.
    - Add the TopologyFactory, a TopologyProvider that merges the
      results of other TopologyProviders.
    - Fix error in NetDevClassBuilder.
    - Add weighted inputs to NewTopologyFactory, to make explicit
      the expected order in which TopologyProviders should be run.
    - Make improvements to debug logging.

Signed-off-by: Kris Jacque <kristin.jacque@intel.com>

### CODE REVIEW

**Comment by mjmac** [comment]
@kjacque Quick update for when you're back at this...

Good news, the CXI devices were detected with the new library and the sysfs provider. Bad news, there seems to be some problem with the NUMA affinity logic.

netdetect net-scan:
```
macdonal@daos2:~> daos_agent net-scan --debug=net
DEBUG 08:11:11.295731 main.go:154: net debug output enabled
DEBUG 08:11:11.459851 netdetect.go:288: 2 NUMA nodes detected with 36 cores per node
DEBUG 08:11:11.481124 netdetect.go:293: network detection, system names: [lo ens785 ens801 ens817f0 ens817f1 bond0], hwloc names [dma0chan0 dma1chan0 dma2chan0 dma3chan0 dma4chan0 dma5chan0 dma6chan0 dma7chan0 sda card0 nvme0n1 ens785 pmem0 dma8chan0 dma9chan0 dma10chan0 dma11chan0 dma12chan0 dma13chan0 dma14chan0 dma15chan0 ens817f0 ens817f1 ens801 nvme1n1 pmem1]
DEBUG 08:11:11.484857 netdetect.go:1231: There are 0 hfi1 devices in the system
DEBUG 08:11:11.488448 netdetect.go:519: scanning sysfs for "cxi0"
DEBUG 08:11:12.494547 netdetect.go:1148: The target device name has been converted to system device name: ens785
DEBUG 08:11:12.494881 netdetect.go:519: scanning sysfs for "cxi1"
DEBUG 08:11:13.749697 netdetect.go:1148: The target device name has been converted to system device name: ens801
---------
localhost
---------

    -------------
    NUMA Socket 0
    -------------

        Provider        Interfaces
        --------        ----------
        ofi+tcp;ofi_rxm ens785
        udp;ofi_rxd     ens785
        udp             ens785
        ofi+tcp         ens785
        ofi+cxi         ens785

    -------------
    NUMA Socket 1
    -------------

        Provider        Interfaces
        --------        ----------
        ofi+cxi         ens801
        ofi+tcp;ofi_rxm ens801
        udp;ofi_rxd     ens801
        udp             ens801
        ofi+tcp         ens801
```

lib/hardware net-scan:
```
macdonal@daos2:~> ~/daos/install/bin/daos_agent net-scan --debug=net
DEBUG 08:13:42.242870 main.go:154: basic debug output enabled
DEBUG 08:13:42.243153 provider.go:52: getting topology with hwloc version 0x20400
DEBUG 08:13:46.274151 fabric.go:626: initialized FabricScanner
DEBUG 08:13:46.277344 provider.go:102: found fabric interface "ens785": ens785 (providers: ofi+tcp;ofi_rxm)
DEBUG 08:13:46.277404 provider.go:102: found fabric interface "ens801": ens801 (providers: ofi+tcp;ofi_rxm)
DEBUG 08:13:46.277439 provider.go:102: found fabric interface "ens785": ens785 (providers: ofi+tcp;ofi_rxm)
DEBUG 08:13:46.277497 provider.go:102: found fabric interface "ens801": ens801 (providers: ofi+tcp;ofi_rxm)
DEBUG 08:13:46.277529 provider.go:102: found fabric interface "bond0": bond0 (providers: ofi+tcp;ofi_rxm)
DEBUG 08:13:46.277577 provider.go:102: found fabric interface "bond0": bond0 (providers: ofi+tcp;ofi_rxm)
DEBUG 08:13:46.277622 provider.go:102: found fabric interface "lo": lo (providers: ofi+tcp;ofi_rxm)
DEBUG 08:13:46.277670 provider.go:102: found fabric interface "lo": lo (providers: ofi+tcp;ofi_rxm)
DEBUG 08:13:46.277708 provider.go:102: found fabric interface "ens785": ens785 (providers: ofi+tcp;ofi_rxm)
DEBUG 08:13:46.277761 provider.go:102: found fabric interface "ens801": ens801 (providers: ofi+tcp;ofi_rxm)
DEBUG 08:13:46.277791 provider.go:102: found fabric interface "ens785": ens785 (providers: ofi+tcp;ofi_rxm)
DEBUG 08:13:46.277840 provider.go:102: found fabric interface "ens801": ens801 (providers: ofi+tcp;ofi_rxm)
DEBUG 08:13:46.277889 provider.go:102: found fabric interface "bond0": bond0 (providers: ofi+tcp;ofi_rxm)
DEBUG 08:13:46.277917 provider.go:102: found fabric interface "bond0": bond0 (providers: ofi+tcp;ofi_rxm)
DEBUG 08:13:46.277946 provider.go:102: found fabric interface "lo": lo (providers: ofi+tcp;ofi_rxm)
DEBUG 08:13:46.277996 provider.go:102: found fabric interface "lo": lo (providers: ofi+tcp;ofi_rxm)
DEBUG 08:13:46.278026 provider.go:102: found fabric interface "ens785": ens785 (providers: ofi+tcp;ofi_rxm)
DEBUG 08:13:46.278071 provider.go:102: found fabric interface "ens801": ens801 (providers: ofi+tcp;ofi_rxm)
DEBUG 08:13:46.278122 provider.go:102: found fabric interface "ens785": ens785 (providers: ofi+tcp;ofi_rxm)
DEBUG 08:13:46.278361 provider.go:102: found fabric interface "ens801": ens801 (providers: ofi+tcp;ofi_rxm)
DEBUG 08:13:46.278402 provider.go:102: found fabric interface "bond0": bond0 (providers: ofi+tcp;ofi_rxm)
DEBUG 08:13:46.278440 provider.go:102: found fabric interface "bond0": bond0 (providers: ofi+tcp;ofi_rxm)
DEBUG 08:13:46.278479 provider.go:102: found fabric interface "lo": lo (providers: ofi+tcp;ofi_rxm)
DEBUG 08:13:46.278507 provider.go:102: found fabric interface "lo": lo (providers: ofi+tcp;ofi_rxm)
DEBUG 08:13:46.278541 provider.go:102: found fabric interface "ens785": ens785 (providers: udp;ofi_rxd)
DEBUG 08:13:46.278623 provider.go:102: found fabric interface "ens801": ens801 (providers: udp;ofi_rxd)
DEBUG 08:13:46.278660 provider.go:102: found fabric interface "ens785": ens785 (providers: udp;ofi_rxd)
DEBUG 08:13:46.278713 provider.go:102: found fabric interface "ens801": ens801 (providers: udp;ofi_rxd)
DEBUG 08:13:46.279146 provider.go:102: found fabric interface "bond0": bond0 (providers: udp;ofi_rxd)
DEBUG 08:13:46.279429 provider.go:102: found fabric interface "bond0": bond0 (providers: udp;ofi_rxd)
DEBUG 08:13:46.279537 provider.go:102: found fabric interface "lo": lo (providers: udp;ofi_rxd)
DEBUG 08:13:46.280362 provider.go:102: found fabric interface "lo": lo (providers: udp;ofi_rxd)
DEBUG 08:13:46.280538 provider.go:102: found fabric interface "ens785": ens785 (providers: udp)
DEBUG 08:13:46.280691 provider.go:102: found fabric interface "ens801": ens801 (providers: udp)
DEBUG 08:13:46.280835 provider.go:102: found fabric interface "ens785": ens785 (providers: udp)
DEBUG 08:13:46.280975 provider.go:102: found fabric interface "ens801": ens801 (providers: udp)
DEBUG 08:13:46.281063 provider.go:102: found fabric interface "bond0": bond0 (providers: udp)
DEBUG 08:13:46.281162 provider.go:102: found fabric interface "bond0": bond0 (providers: udp)
DEBUG 08:13:46.281242 provider.go:102: found fabric interface "lo": lo (providers: udp)
DEBUG 08:13:46.281320 provider.go:102: found fabric interface "lo": lo (providers: udp)
DEBUG 08:13:46.281467 provider.go:102: found fabric interface "ens785": ens785 (providers: ofi+tcp)
DEBUG 08:13:46.281611 provider.go:102: found fabric interface "ens801": ens801 (providers: ofi+tcp)
DEBUG 08:13:46.281697 provider.go:102: found fabric interface "ens785": ens785 (providers: ofi+tcp)
DEBUG 08:13:46.281781 provider.go:102: found fabric interface "ens801": ens801 (providers: ofi+tcp)
DEBUG 08:13:46.281923 provider.go:102: found fabric interface "bond0": bond0 (providers: ofi+tcp)
DEBUG 08:13:46.282002 provider.go:102: found fabric interface "bond0": bond0 (providers: ofi+tcp)
DEBUG 08:13:46.282073 provider.go:102: found fabric interface "lo": lo (providers: ofi+tcp)
DEBUG 08:13:46.282144 provider.go:102: found fabric interface "lo": lo (providers: ofi+tcp)
DEBUG 08:13:46.282285 provider.go:102: found fabric interface "cxi0": cxi0 (interface: cxi0) (providers: ofi+cxi)
DEBUG 08:13:46.282426 provider.go:102: found fabric interface "cxi1": cxi1 (interface: cxi1) (providers: ofi+cxi)
DEBUG 08:13:46.282562 fabric.go:361: updating fabric interface "bond0"
DEBUG 08:13:46.282649 fabric.go:361: updating fabric interface "cxi0"
DEBUG 08:13:46.282716 fabric.go:361: updating fabric interface "cxi1"
DEBUG 08:13:46.282780 fabric.go:361: updating fabric interface "ens785"
DEBUG 08:13:46.282841 fabric.go:361: updating fabric interface "ens801"
DEBUG 08:13:46.282896 fabric.go:361: updating fabric interface "lo"
DEBUG 08:13:46.283167 fabric.go:531: open /sys/class/net/cxi0/type: no such file or directory
DEBUG 08:13:46.283257 fabric.go:531: open /sys/class/net/cxi1/type: no such file or directory
DEBUG 08:13:46.283566 fabric.go:415: fabric interface "bond0" not found in topology
DEBUG 08:13:46.283718 fabric.go:486: fabric interface "bond0" not found in topology
---------
localhost
---------

    -------------
    NUMA Socket 0
    -------------

        Provider        Interfaces
        --------        ----------
        udp             bond0, ens785
        udp;ofi_rxd     bond0, ens785
        ofi+cxi         ens785, ens801
        ofi+tcp         bond0, ens785
        ofi+tcp;ofi_rxm bond0, ens785

    -------------
    NUMA Socket 1
    -------------

        Provider        Interfaces
        --------        ----------
        ofi+tcp         ens801
        ofi+tcp;ofi_rxm ens801
        udp             ens801
        udp;ofi_rxd     ens801
```

dump-topology:
```
macdonal@daos2:~> ~/daos/install/bin/daos_agent dump-topology
NUMA Node 0
  CPU cores: 0-35
  PCI devices:
    0000:88:00.0
      0000:88:00.0 cxi1 (OFI domain)
      0000:88:00.0 cxi1 (OFI domain)
      0000:88:00.0 ens801 (network interface)
      0000:88:00.0 ptp1 (OFI domain)
    0000:07:00.0
      0000:07:00.0 ens785 (network interface) @ 31.51 GB/s
      0000:07:00.0 cxi0 (OFI domain)
      0000:07:00.0 cxi0 (OFI domain)
      0000:07:00.0 ptp0 (OFI domain)
    0000:82:00.0
      0000:82:00.0 ens817f0 (network interface)
      0000:82:00.0 device (OFI domain)
      0000:82:00.0 ptp2 (OFI domain)
      0000:82:00.0 lower_ens817f0 (OFI domain)
    0000:82:00.1
      0000:82:00.1 ens817f1 (network interface)
      0000:82:00.1 device (OFI domain)
  PCI buses:
    0000:[00-02]
    0000:03
    0000:[04-05]
    0000:[06-08]
    0000:[79-7d]
    0000:7e
    0000:7f
NUMA Node 1
  CPU cores: 36-71
  PCI devices:
    0000:82:00.0
      0000:82:00.0 ens817f0 (network interface) @ 3.94 GB/s
    0000:82:00.1
      0000:82:00.1 ens817f1 (network interface) @ 3.94 GB/s
    0000:88:00.0
      0000:88:00.0 ens801 (network interface) @ 31.51 GB/s
  PCI buses:
    0000:80
    0000:[81-83]
    0000:[87-89]
    0000:[f4-f5]
    0000:[f6-fa]
    0000:fe
    0000:ff
```

**Comment by mjmac** [comment]
Interestingly, when I run dump-topology with a build from current master, it all looks fine (but no CXI devices, as expected):
```
macdonal@daos2:~> ~/daos/install/bin/daos_agent dump-topology
NUMA Node 0
  CPU cores: 0-35
  PCI devices:
    0000:07:00.0
      0000:07:00.0 ens785 (network interface) @ 31.51 GB/s
  PCI buses:
    0000:[00-02]
    0000:03
    0000:[04-05]
    0000:[06-08]
    0000:[79-7d]
    0000:7e
    0000:7f
NUMA Node 1
  CPU cores: 36-71
  PCI devices:
    0000:82:00.1
      0000:82:00.1 ens817f1 (network interface) @ 3.94 GB/s
    0000:88:00.0
      0000:88:00.0 ens801 (network interface) @ 31.51 GB/s
    0000:82:00.0
      0000:82:00.0 ens817f0 (network interface) @ 3.94 GB/s
  PCI buses:
    0000:80
    0000:[81-83]
    0000:[87-89]
    0000:[f4-f5]
    0000:[f6-fa]
    0000:fe
    0000:ff
```

File: src/control/lib/hardware/sysfs/provider.go [hunk 1]
```diff
@@ -35,17 +42,95 @@ func (s *Provider) getRoot() string {
...
 	if err != nil {
 		return 0, err
 	}
 
 	res, err := strconv.Atoi(strings.TrimSpace(string(devClass)))
 	return hardware.NetDevClass(res), err
 }
+
+// GetTopology builds a minimal topology of network devices from the contents of sysfs.
+func (s *Provider) GetTopology(ctx context.Context) (*hardware.Topology, error) {
```
Comment (mjmac) [inline src/control/lib/hardware/sysfs/provider.go:67]: Don't want to over-engineer, but I do think we probably want to use this stuff for more than finding network devices (e.g. storage, etc), so I wonder if maybe the interface should accept a config struct or maybe a bitmask which at a minimum defines which types of devices will be included in the output topology. You could then break the logic below out into smaller helper functions for finding various device types.

File: src/control/lib/hardware/topology.go [hunk 2]
```diff
@@ -68,6 +68,101 @@ func (t *Topology) NumCoresPerNUMA() int {
 	return 0
 }
 
+// AddDevice adds a device to the topology.
+func (t *Topology) AddDevice(numaID uint, device *PCIDevice) {
+	if t == nil || device == nil {
+		return
+	}
```
Comment (mjmac) [inline src/control/lib/hardware/topology.go:75]: Per our conversation in the other PR, I'm a little concerned that this pattern of ignoring programming errors may lead to confusing behavior or subtle bugs, particularly in light of the fact that this is all new code being used to explore new hardware topologies. I am sympathetic to the position that adding error handing on the part of the caller is less elegant, and I'd be open to the idea of using `panic()` in some of these cases if we're pretty sure that that these are "can't happen" situations.

File: src/control/lib/hardware/topology.go [hunk 3]
```diff
@@ -262,3 +357,29 @@ func (t DeviceType) String() string {
...
+// GetTopology gets a merged master topology from all the topology providers.
+func (tf *TopologyFactory) GetTopology(ctx context.Context) (*Topology, error) {
+	newTopo := &Topology{}
+	for _, prov := range tf.providers {
+		topo, err := prov.GetTopology(ctx)
+		if err != nil {
+			return nil, err
+		}
+		newTopo.Merge(topo)
+	}
```
Comment (mjmac) [inline src/control/lib/hardware/topology.go:376]: I think this is fine for now, and perhaps always will be. It occurs to me to wonder whether or not we need to somehow decide which topologies should be grabbed first based on some kind of weighting mechanism, though. As it is, if there is an ordering dependency, it requires the person writing the code to understand it and make sure that the providers list is set up correctly.

File: src/control/lib/hardware/topology.go (hunk 2 above)
Comment (kjacque) [reply src/control/lib/hardware/topology.go:75]: My thought was that the action is meaningless on a nil caller or nil input. In theory it shouldn't happen, but at this low a level it's always a little hard to guarantee caller behavior. Maybe it should just return an error, even if that looks ugly.

File: src/control/lib/hardware/topology.go (hunk 3 above)
Comment (kjacque) [reply src/control/lib/hardware/topology.go:279]: I added a weighting mechanism to the inputs for the NewTopologyFactory method.

**Review by mjmac — APPROVED** [review]

**Review by tanabarr — DISMISSED** [review]

File: src/control/lib/hardware/topology.go (hunk 2 above)
Comment (kjacque) [reply src/control/lib/hardware/topology.go]: I thought I'd save this for a follow-on, but I ended up taking your advice to solve a bug involving a corner case for the Topology merge.  Lesson learned.

**Review by tanabarr — APPROVED** [review]

**Review by tanabarr — APPROVED** [review]

File: src/control/lib/hardware/sysfs/provider.go [hunk 4]
```diff
@@ -191,10 +190,54 @@ func (s *Provider) getPCIAddress(path string) (*hardware.PCIAddress, error) {
...
+			return pciAddr, nil
+		}
+
+		pciPath = filepath.Dir(pciPath)
+	}
+
+	return nil, errors.Errorf("unable to parse PCI address from %q", path)
+}
+
+// GetFabricInterfaces harvests the CXI fabric interfaces from sysfs.
```
Comment (mjmac) [inline src/control/lib/hardware/sysfs/provider.go:206]: It's true that for the moment, we're only harvesting CXI interfaces using sysfs, but I'm not sure that we should make the API documentation specific to that use case.

**Review by mjmac — APPROVED** [review]
LGTM; I like the increased test coverage. Does this also address the problem that @jolivier23 found on his VM?

**Comment by kjacque** [comment]
> Does this also address the problem that @jolivier23 found on his VM?

It does.

File: src/control/lib/hardware/sysfs/provider.go [hunk 5]
```diff
@@ -191,10 +190,54 @@ func (s *Provider) getPCIAddress(path string) (*hardware.PCIAddress, error) {
...
+	cxiDevs, err := ioutil.ReadDir(s.sysPath("class", "cxi"))
+	if os.IsNotExist(err) {
+		s.log.Debugf("no cxi subsystem in sysfs")
+		return []*hardware.FabricInterface{}, nil
+	} else if err != nil {
+		return nil, err
+	}
+
+	if len(cxiDevs) == 0 {
+		s.log.Debugf("no cxi devices in sysfs")
```
[the remark this answers is not in the document]
Comment (kjacque) [reply src/control/lib/hardware/sysfs/provider.go:230]: That's fair. I had some interesting problems while adding the new functionality and while debugging Jeff's issue, so I ended up adjusting the debug output all around the network stuff.

File: src/control/lib/hardware/sysfs/provider.go (hunk 4 above)
Comment (kjacque) [reply src/control/lib/hardware/sysfs/provider.go:206]: Good point. I can adjust this in my next PR, which is in the works at the moment.

**Pull request merged**